cmd/xl-packs.go

```diff
@@ -0,0 +1,437 @@
+/*
+ * Minio Cloud Storage, (C) 2017 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+import (
+	"io"
+	"sort"
+	"strings"
+
+	"github.com/minio/minio/pkg/errors"
+	"github.com/minio/minio/pkg/hash"
+	"golang.org/x/sync/errgroup"
+)
+
+// xlPacks implements Objectlayer combining a static list of erasure coded
+// object layers. NOTE: There is no dynamic scaling allowed or intended in
+// current design.
+type xlPacks struct {
+	layers []*xlObjects
+
+	// Pack level listObjects pool management.
+	listPool *treeWalkPool
+}
+
+// Initialize new set of erasure coded packs.
+func newXLPacks(layers []*xlObjects) ObjectLayer {
+	return &xlPacks{
+		layers:   layers,
+		listPool: newTreeWalkPool(globalLookupTimeout),
+	}
+}
+
+// StorageInfo - combines output of StorageInfo across all erasure
+// coded object layers.
+func (s xlPacks) StorageInfo() StorageInfo {
+	var info StorageInfo
+	info.Backend.Type = Erasure
+	for _, layer := range s.layers {
+		linfo := layer.StorageInfo()
+		info.Total = info.Total + linfo.Total
+		info.Free = info.Free + linfo.Free
+		info.Backend.OnlineDisks = info.Backend.OnlineDisks + linfo.Backend.OnlineDisks
+		info.Backend.OfflineDisks = info.Backend.OfflineDisks + linfo.Backend.OfflineDisks
+	}
+	return info
+}
+
+func (s xlPacks) Shutdown() error {
+	var g errgroup.Group
+
+	for _, layer := range s.layers {
+		layer := layer // https://golang.org/doc/faq#closures_and_goroutines
+		g.Go(func() error {
+			return layer.Shutdown()
+		})
+	}
+
+	return g.Wait()
+}
+
+func (s xlPacks) MakeBucketWithLocation(bucket, location string) error {
+	var g errgroup.Group
+
+	for _, layer := range s.layers {
+		layer := layer // https://golang.org/doc/faq#closures_and_goroutines
+		g.Go(func() error {
+			return layer.MakeBucketWithLocation(bucket, location)
+		})
+	}
+
+	// Wait for all makebucket to finish.
+	return g.Wait()
```

Comment:
What do we do when bucket creation fails in, for example, just one slot? (Eg. because one or more servers are down for that particular pack only).
If we return overall success then we should be resistent to a bucket not being available for a slot (depending on the object's name... ).